src/main/java/ch/njol/skript/lang/SkriptParser.java

```diff
@@ -263,7 +263,30 @@ private <T extends SyntaxElement> T parse(Iterator<? extends SyntaxElementInfo<?
 								x = x2;
 							}
 							T t = info.c.newInstance();
-							if (t.init(res.exprs, i, getParser().getHasDelayBefore(), res)) {
+							ParserInstance parser = getParser();
+							List<Class<? extends Structure>> usableStructures = t.getUsableStructures();
+							if (!usableStructures.isEmpty()) {
+								if (parser.getCurrentStructure() == null) {
+									Skript.error("'" + res.expr + "' cannot be used outside of the allowed structures.");
+									continue;
+								} else if (usableStructures.stream().noneMatch(parser::isCurrentStructure)) {
+									Skript.error("'" + res.expr + "' cannot be used in " + Utils.a(parser.getCurrentStructure().toString(null, false)) + " structure.");
+									continue;
+								}
+							}
+							List<Class<? extends Section>> usableSections = t.getUsableSections();
+							if (!usableSections.isEmpty()) {
+								List<TriggerSection> sections = parser.getCurrentSections();
+								TriggerSection current = sections.isEmpty() ? null : sections.get(sections.size() - 1);
+								if (sections.isEmpty()) {
+									Skript.error("'" + res.expr + "' cannot be used outside of the allowed sections.");
+									continue;
+								} else if (usableSections.stream().noneMatch(parser::isCurrentSection)) {
+									Skript.error("'" + res.expr + "' cannot be used in '" + Utils.a(current.toString(null, false)) + "' section.");
```

Comment:
Does this variable need to be here? Can't it be included inside the if condition it's only used in?
```suggestion
								if (sections.isEmpty()) {
									Skript.error("'" + res.expr + "' cannot be used outside of the allowed sections.");
									continue;
								} else if (usableSections.stream().noneMatch(parser::isCurrentSection)) {
									Skript.error("'" + res.expr + "' cannot be used in '" + Utils.a(sections.get(sections.size() - 1).toString(null, false)) + "' section.");
```